core/domain/exp_domain.py

```diff
@@ -2964,6 +2967,82 @@ def delete_state(self, state_name):
 
         del self.states[state_name]
 
+
+    def get_state_names_mapping(self, change_list):
+        """Creates a mapping from the current state names of the exploration
+        to their corresponding state names of the older version.
+
+        Args:
+            change_list: list(dict). A list of changes introduced in this
+                commit.
+
+        Returns:
+            dict. The new_to_old_state_names mapping dict.
+        """
+        old_to_new_state_names = {}
+
+        for state_name in self.states:
+            old_to_new_state_names[state_name] = state_name
+
+        for change_dict in reversed(change_list):
+            if change_dict['cmd'] == CMD_RENAME_STATE:
+                old_to_new_state_names[change_dict['old_state_name']] = (
+                    old_to_new_state_names.pop(change_dict['new_state_name']))
+
+        new_to_old_state_names = {
+            value: key for key, value in old_to_new_state_names.iteritems()
+        }
+        return new_to_old_state_names
+
+
+    def get_trainable_states_dict(self, old_states, new_to_old_state_names):
+        """Retrieves the state names of all trainable states in an exploration
+        segregated into state names with changed and unchanged answer groups.
+
+        Args:
+            old_states: dict. Dictionary containing all State domain objects.
+            new_to_old_state_names: dict. A mapping from current state names to
+                state names from previous version of the exploration.
+
+        Returns:
+            dict. The trainable states dict. This dict has three keys
+                representing state names with changed answer groups and
+                unchanged answer groups respectively.
+        """
+        trainable_states_dict = {
+            'state_names_with_changed_answer_groups': [],
+            'state_names_with_unchanged_answer_groups': []
+        }
+        new_states = self.states
+
+        for state_name in new_states:
+            new_state = new_states[state_name]
+            if not new_state.can_undergo_classification():
+                continue
+
+            old_state_name = new_to_old_state_names[state_name]
+
+            # The case where a new state is added. When this happens, there
+            # won't be a corresponding state name in the older state dict.
+            if old_state_name not in old_states:
```

Comment:
hmm this looks a bit iffy while reading. Anyone reading this would think that the mapping `new_to_old_state_names`  is created using new_states and old_states, and, when you do `if old_state_name not in old_states`, people would be terribly confused. We need to be clear in the comments about what is happening here.